### PR TITLE
schemadiff: normalize index option value (string)

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

With this PR `schemadiff` normalizes (to lower case) the value of an index option. For example: 

`fulltext key name_ft(name) WITH PARSER NGRAM` is normalized into
`fulltext key name_ft(name) WITH PARSER ngram`

## Related Issue(s)

Tracking: https://github.com/vitessio/vitess/issues/10203

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [ ] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/table.go [hunk 1]
```diff
@@ -482,6 +482,7 @@ func (c *CreateTableEntity) normalizeIndexOptions() {
 		idx.Info.Type = strings.ToLower(idx.Info.Type)
 		for _, opt := range idx.Options {
 			opt.Name = strings.ToLower(opt.Name)
+			opt.String = strings.ToLower(opt.String)
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:485]: Hmm, would this always lowercase things if you have for example a comment set? It seems like we should not normalize that then. I wonder if we should make this specific only to `WITH PARSER`?

File: go/vt/schemadiff/table_test.go [hunk 2]
```diff
@@ -438,6 +438,20 @@ func TestCreateTableDiff(t *testing.T) {
 			diff:  "alter table t1 add fulltext key name_ft (`name`)",
 			cdiff: "ALTER TABLE `t1` ADD FULLTEXT KEY `name_ft` (`name`)",
 		},
+		{
+			name:  "add one fulltext key with explicit parser",
+			from:  "create table t1 (id int primary key, name tinytext not null)",
+			to:    "create table t1 (id int primary key, name tinytext not null, fulltext key name_ft(name) with parser ngram)",
+			diff:  "alter table t1 add fulltext key name_ft (`name`) with parser ngram",
+			cdiff: "ALTER TABLE `t1` ADD FULLTEXT KEY `name_ft` (`name`) WITH PARSER NGRAM",
```
Comment (dbussink) [inline go/vt/schemadiff/table_test.go:446]: I think we also want to / need to change the canonical output format of this. It looks like the parser formatter always upcases this. I think we should add something like the following as well:

```patch
diff --git i/go/vt/sqlparser/ast_format.go w/go/vt/sqlparser/ast_format.go
index e94a1b24ab..65756a46a9 100644
--- i/go/vt/sqlparser/ast_format.go
+++ w/go/vt/sqlparser/ast_format.go
@@ -826,7 +826,7 @@ func (idx *IndexDefinition) Format(buf *TrackedBuffer) {
        for _, opt := range idx.Options {
                buf.astPrintf(idx, " %s", opt.Name)
                if opt.String != "" {
-                       buf.astPrintf(idx, " %s", opt.String)
+                       buf.astPrintf(idx, " %#s", opt.String)
                } else if opt.Value != nil {
                        buf.astPrintf(idx, " %v", opt.Value)
                }
diff --git i/go/vt/sqlparser/tracked_buffer_test.go w/go/vt/sqlparser/tracked_buffer_test.go
index 02ea192a5d..614a215c0e 100644
--- i/go/vt/sqlparser/tracked_buffer_test.go
+++ w/go/vt/sqlparser/tracked_buffer_test.go
@@ -220,6 +220,10 @@ func TestCanonicalOutput(t *testing.T) {
                        "select char(77, 121, 83, 81, '76' using utf8mb4) from dual",
                        "SELECT CHAR(77, 121, 83, 81, '76' USING utf8mb4) FROM `dual`",
                },
+               {
+                       "create table t1 (id int primary key, name tinytext not null, fulltext key name_ft(name) with parser ngram)",
+                       "CREATE TABLE `t1` (\n\t`id` int PRIMARY KEY,\n\t`name` tinytext NOT NULL,\n\tFULLTEXT KEY `name_ft` (`name`) WITH PARSER ngram\n)",
+               },
        }
 
        for _, tc := range testcases {
```

File: go/vt/schemadiff/table.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:485]: Per our parser:

https://github.com/vitessio/vitess/blob/1ace3b47046a40e4c43b1ee0d2088db01f35ef0d/go/vt/sqlparser/sql.y#L2299-L2332

It looks like the `String` value is only ever used (at this time?) in a `WITH PARSER` clause. No other index clause/option utilizes it.

Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:485]: hmmm also here:

https://github.com/vitessio/vitess/blob/1ace3b47046a40e4c43b1ee0d2088db01f35ef0d/go/vt/sqlparser/sql.y#L7240-L7244

`USING ...`, so this affects e.g. `USING HASH` -> `USING hash`

File: go/vt/schemadiff/table_test.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table_test.go]: added

**Comment by shlomi-noach** [comment]
Normalization now takes place in both `ast_format` and in `schemadiff`. Only two syntax are affected by this chane:

- `USING ...` for "normal" index types (`USING HASH`, `USING BTREE`)
- `WITH PARSER ...` for fulltext index types (`USING PARSER NGRAM`)

`schemadiff` tests adapted. `sqlparser` tests unchanged and all good.

File: go/vt/schemadiff/table.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/schemadiff/table.go:485]: I don't see any issues with lower-casing `USING` or `WITH PARSER` values

**Review by GuptaManan100 — APPROVED** [review]

**Pull request merged**